### PR TITLE
docs: Add data definition document

### LINKED ISSUE #2538: Review schema vs collection documentation
Public documentation, both in the Go codebase (i.e. `client` package), and the docs repo.  Consider adding a .md file to the defra repo.

Particularly regarding global vs local fields.

#1386 and #1385 might be handy here too.

### PR DESCRIPTION
**Pull request by AndrewSisley**
## Relevant issue(s)

Resolves #2538

## Description

Adds data definition document, documenting how data is defined.

It is targeted at us, and anyone looking to embed defradb in their application.

### CODE REVIEW

File: client/data_definition.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+Splitting local elements out from the global ones allows some local customization to the way data is organized within any given node.  It also minimizes the amount of 'stuff' that must be kept consistent across the decentralized network in order to have a well behaving database.
+
+Local data definitions are always defined on the [collection](#collections).
+
+Examples include indexes, field IDs, and [lens transforms](https://docs.source.network/defradb/guides/schema-migration).
+
+## Global definitions
+
+Global definitions are consistent across all nodes in the decentralized network of nodes in the database. This is enforced by the use of things like CIDs for schema versions.  If a global definition was to differ across nodes, the different variations will be treated as a completely different definitions.
```
Comment (nasdf) [inline client/data_definition.md:17]: suggestion: Global definitions are consistent across all nodes in the decentralized network ~of nodes in the database~

File: client/data_definition.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+Collections are defined by the `CollectionDescription` struct.  This can be mutated via the `PatchCollection` function.
+
+A collection will always have a [global](#global-definitions) shape defined by a single [schema](#schemas) version.
+
+### Versions
+
+`CollectionDescription` instances may be active or inactive.  Inactive `CollectionDescription`s will not have a name, and cannot be queried.
+
+When a new [schema](#schemas) version is created for a schema that has a collection defined for it, a new `CollectionDescription` instance will be created and linked to the new schema version.  The new `CollectionDescription` instance will share the same root ID as the previous, and may be active or inactive depending on what arguments the user defining the new schema specified.
```
Comment (nasdf) [inline client/data_definition.md:35]: suggestion: When a new [schema](#schemas) version is created ~for a schema that~ *and* has a collection defined for it, a new `CollectionDescription` instance will be created and linked to the new schema version.

**Review by nasdf — APPROVED** [review]
Great job! Documentation like this so helpful. I hope to see more of this in the future.

**Review by shahzadlone — APPROVED** [review]
Cheers! Huge praise on this helpful documentation.

File: client/data_definition.md (hunk 1 above)
Comment (AndrewSisley) [reply client/data_definition.md:17]: Will do :) I wasn't terribly confident writing that last bit out, so I'm very glad you spoke up :)

- [ ] Drop extra words

File: client/data_definition.md (hunk 2 above)
Comment (AndrewSisley) [reply client/data_definition.md:35]: Yeah, I think I agree - will do, thanks Keenan :)

- [ ] Remove extra words

Comment (AndrewSisley) [reply client/data_definition.md]: praise: The italics on `and` in your suggestion saved me a bit of hassle, thanks for bothering to add that :)

**Pull request merged**